crates/sui-config/src/node.rs

```diff
@@ -388,8 +393,144 @@ enum GenesisLocation {
+            keypair: cell,
+        }
+    }
+
+    pub fn keypair(&self) -> &SuiKeyPair {
+        self.keypair
+            .get_or_init(|| match &self.location {
+                KeyPairLocation::InPlace { value } => value.clone(),
+                KeyPairLocation::File { path } => {
+                    // OK to unwrap panic because authority should not start without all keypairs loaded.
+                    Arc::new(
+                        read_keypair_from_file(path)
+                            .unwrap_or_else(|_| panic!("Invalid keypair file")),
+                    )
+                }
+            })
+            .as_ref()
+    }
+}
+
+/// Wrapper struct for AuthorityKeyPair that can be deserialized from a file path.
+#[derive(Clone, Debug, PartialEq, Eq, Deserialize, Serialize)]
+pub struct AuthorityKeyPairWithPath {
+    #[serde(flatten)]
+    location: AuthorityKeyPairLocation,
+
+    #[serde(skip)]
+    keypair: OnceCell<Arc<AuthorityKeyPair>>,
+}
+
+#[derive(Debug, Clone, PartialEq, Deserialize, Serialize, Eq)]
+#[serde_as]
+#[serde(untagged)]
+enum AuthorityKeyPairLocation {
+    InPlace { value: Arc<AuthorityKeyPair> },
+    File { path: PathBuf },
+}
+
+impl AuthorityKeyPairWithPath {
+    pub fn new(kp: AuthorityKeyPair) -> Self {
+        let cell: OnceCell<Arc<AuthorityKeyPair>> = OnceCell::new();
+        let arc_kp = Arc::new(kp);
+        // OK to unwrap panic because authority should not start without all keypairs loaded.
+        cell.set(arc_kp.clone())
+            .expect("Failed to set authority keypair");
+        Self {
+            location: AuthorityKeyPairLocation::InPlace { value: arc_kp },
+            keypair: cell,
+        }
+    }
+
+    pub fn new_from_path(path: PathBuf) -> Self {
+        let cell: OnceCell<Arc<AuthorityKeyPair>> = OnceCell::new();
+        // OK to unwrap panic because authority should not start without all keypairs loaded.
+        cell.set(Arc::new(
+            read_authority_keypair_from_file(&path)
+                .unwrap_or_else(|_| panic!("Invalid authority keypair file at path {:?}", &path)),
+        ))
+        .expect("Failed to set authority keypair");
+        Self {
+            location: AuthorityKeyPairLocation::File { path },
+            keypair: cell,
+        }
+    }
+
+    pub fn authority_keypair(&self) -> &AuthorityKeyPair {
+        self.keypair
+            .get_or_init(|| match &self.location {
+                AuthorityKeyPairLocation::InPlace { value } => value.clone(),
+                AuthorityKeyPairLocation::File { path } => {
+                    // OK to unwrap panic because authority should not start without all keypairs loaded.
+                    Arc::new(
+                        read_authority_keypair_from_file(path).unwrap_or_else(|_| {
+                            panic!("Invalid authority keypair file {:?}", &path)
+                        }),
+                    )
+                }
+            })
+            .as_ref()
+    }
+}
+
```

Comment:
Currently, BLS is not included in the `enum SuiKeyPair` and hence we have some code dup on `KeyPairWithPath` vs `AuthorityKeyPairWithPath`. I will follow up a cleanup PR to deprecate `AuthorityKeyPair` fully (it is not trivial change to add it)